src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdSlotBuilder.java

```diff
@@ -0,0 +1,165 @@
+            };
+        } else if (imp.getAudio() != null) {
+            throw new PreBidException(
+                    "check openrtb format: request has audio, not currently supported");
+        } else {
+            throw new PreBidException(
+                    "check openrtb format: please choose one of our supported type banner, native, or video");
+        }
+
+        return adSlot.toBuilder()
+                .slotId(impExt.getSlotId())
+                .adType(adsType.getType())
+                //todo: IsTestAuthorization looks unclear, also BidRequest also has test object, what's better to use?
+                //todo: does make sense to make it more robust (at least case insensitive)
+                .test(TEST_AUTH_ENABLED.equals(impExt.getIsTestAuthorization()) ? 1 : 0)
+                .build();
+    }
+
+    private AdSlot30 makeVideoAdSlot(Video video) {
+        return HuaweiUtils.isFormatDefined(video.getW(), video.getH())
+                ? AdSlot30.builder().w(video.getW()).h(video.getH()).build()
+                : AdSlot30.builder().build();
+    }
+
+    private AdSlot30 makeRollVideoAdSlot(Video video) {
+        final Integer maxDuration = video.getMaxduration();
+        if (maxDuration == null || maxDuration == 0) {
+            throw new PreBidException("extract openrtb video failed: MaxDuration is empty "
+                    + "when huaweiads adtype is roll.");
+        }
+
+        return makeVideoAdSlot(video).toBuilder()
+                .totalDuration(maxDuration)
+                .build();
+    }
+
+    private AdSlot30 makeNativeAdSlot(Native xNative) {
+        final AdSlot30.AdSlot30Builder builder = AdSlot30.builder();
+        final List<Asset> assets = parseNativeRequestAssets(xNative);
+        // only compute the main image number
+        final List<ImageObject> images = assets.stream().map(Asset::getImg).filter(Objects::nonNull)
+                .filter(img -> IMAGE_ASSET_TYPE_MAIN.equals(img.getType()))
+                .toList();
+        // every image has the same W, H.
+        //todo: make sense to try to take format from all the image types as well if main ones are absent
+        images.forEach(image -> {
+            if (HuaweiUtils.isFormatDefined(image.getW(), image.getH())) {
+                builder.w(image.getW()).h(image.getH());
+            } else if (HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                builder.w(image.getWmin()).h(image.getHmin());
+            }
+        });
+
+        // Only one of the {title,img,video,data} objects should be present in each object.
+        final long numVideo = assets.stream().map(Asset::getVideo).filter(Objects::nonNull).count();
+        final long numImage = images.size();
+        builder.detailedCreativeTypeList(makeDetailedCreativeTypeList(numVideo, numImage));
+
+        return builder.build();
+    }
```

Comment:
Can be optimized:
```
    private AdSlot30 makeNativeAdSlot(Native xNative) {
        final List<Asset> assets = parseNativeRequestAssets(xNative);
        // only compute the main image number
        final List<ImageObject> images = assets.stream().map(Asset::getImg).filter(Objects::nonNull)
                .filter(img -> IMAGE_ASSET_TYPE_MAIN.equals(img.getType()))
                .toList();

        // Only one of the {title,img,video,data} objects should be present in each object.
        final int numImage = images.size();
        final long numVideo = assets.stream().map(Asset::getVideo).filter(Objects::nonNull).count();

        final ImageObject lastValid = IntStream.range(0, numImage)
                .mapToObj(i -> images.get(numImage - i - 1))
                .filter(image -> HuaweiUtils.isFormatDefined(image.getW(), image.getH())
                        || HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin()))
                .findFirst()
                .orElse(null);

        return AdSlot30.builder()
                .w(lastValid != null ? lastValid.getW() : null)
                .h(lastValid != null ? lastValid.getH() : null)
                .detailedCreativeTypeList(makeDetailedCreativeTypeList(numVideo, numImage))
                .build();
    }
```

Reply:
this part has been changed in PBS Go recently, so the comment is irrelevant now, please double-check